src/python/pants/ox.py

```diff
@@ -0,0 +1,160 @@
+        _logger.info("Pants is running as a PyOxidizer binary.")
+
+
+# Provide the `is_oxidized` symbol, to allow for workarounds in Pants code where we use things
+# that don't work under PyOxidizer's custom importer. `oxidized_importer` is only accessible
+# in Pants under PyOxidizer, so an import failure will occur if we're not oxidized.
+try:
+    import oxidized_importer  # type: ignore # pants: no-infer-dep # noqa: F401
+
+    is_oxidized = True
+except ModuleNotFoundError:
+    is_oxidized = False
+
+
+if is_oxidized and not sys.argv[0]:
+    # A not insignificant amount of Pants code relies on `sys.argv[0]`, which is modified in an
+    # invalid way by python's `pymain_run_module` support. For our purposes, the executable
+    # distribution is the correct `argv[0]`.
+    # See https://github.com/indygreg/PyOxidizer/issues/307
+    sys.argv[0] = sys.executable
+
+
+def pex_main() -> bool:
+    """Detect whether some external process is trying to invoke this binary as Pex.
+
+    When bootstrapping a plugin venv, Pants will reinvoke Pex in the binary that Pants was loaded
+    with. Pex will subsequently run a bunch of Python processes in order to run pip, venv, and
+    `python -c` invocations. This looks for certain command line switches, and attempt to invoke the
+    relevant modules that those switches indicate.
+    """
+
+    if len(sys.argv) < 2:
+        return False
+
+    if sys.argv[1] == "./pex":
+        _run_as_pex()
+        return True
+
+    if sys.argv[1] == "-sE" and sys.argv[2].endswith("/pex"):
+        _run_pex_venv()
+        return True
+
+    if len(sys.argv) == 4 and sys.argv[1:3] == ["-s", "-c"]:
+        _run_as_dash_c()
+        return True
+
+    if ("-m", "venv") in zip(sys.argv, sys.argv[1:]):
+        _run_as_venv()
+        return True
+
+    return False
+
+
+@contextlib.contextmanager
+def traditional_import_machinery():
+    # pex relies heavily on `__file__`, which the Oxidized importer does not
+    # believe in. This reinstates the default Python import machinery before
+    # loading and running `pex`, but keeps the PyOxidizer machinery at lowest
+    # priority, so we can still load interned `.py` sources (e.g. the stdlib)
+
+    old_sys_meta_path = sys.meta_path
+    old_sys_path_hooks = sys.path_hooks
+
+    if is_oxidized:
+        sys.meta_path = [
+            importlib.machinery.BuiltinImporter,
+            importlib.machinery.FrozenImporter,
+            importlib.machinery.PathFinder,
+        ] + sys.meta_path
+        sys.path_hooks = [
+            zipimport.zipimporter,
+            importlib.machinery.FileFinder.path_hook(
+                (
+                    importlib.machinery.ExtensionFileLoader,
+                    [".cpython-39-darwin.so", ".abi3.so", ".so"],
+                ),
+                (importlib.machinery.SourceFileLoader, [".py"]),
+                (importlib.machinery.SourcelessFileLoader, [".pyc"]),
+            ),
+        ] + sys.path_hooks
+
+    yield
```

Comment:
This should use `try/finally`.

Reply:
good catch, thanks. (I feel like I recently spent several minutes talking about exactly this?)